src/Stratis.Bitcoin.Features.WatchOnlyWallet/WatchOnlyWalletManager.cs

```diff
@@ -155,11 +231,75 @@ public void ProcessTransaction(Transaction transaction, Block block = null)
                         }
+
+                        // Update the lookup cache with the new transaction information.
+                        this.txLookup.Remove(existingTransaction.Id);
+                        this.txLookup.Add(existingTransaction.Id, existingTransaction);
                     }
 
                     this.SaveWatchOnlyWallet();
                 }
             }
+
+            this.Wallet.WatchedTransactions.TryGetValue(transactionHash.ToString(), out TransactionData existingWatchedTransaction);
+
+            if (existingWatchedTransaction != null)
+            {
+                // The transaction was previously stored, in an unconfirmed state.
+                // So now update the block hash and Merkle proof since it has
+                // appeared in a block.
+                existingWatchedTransaction.BlockHash = block.GetHash();
+                existingWatchedTransaction.MerkleProof = new MerkleBlock(block, new[] {transaction.GetHash()}).PartialMerkleTree;
+
+                // Update the lookup cache with the new transaction information.
+                this.txLookup.Remove(existingWatchedTransaction.Id);
+                this.txLookup.Add(existingWatchedTransaction.Id, existingWatchedTransaction);
+
+                this.SaveWatchOnlyWallet();
+            }
+        }
+
+        /// <summary>
+        /// Populate the transaction lookup dictionary with the current
+        /// contents of the watch only wallet's watched addresses and
+        /// transactions.
+        /// </summary>
+        private void LoadTransactionLookup()
+        {
+            var lookup = new Dictionary<uint256, TransactionData>();
+            var watchOnlyWallet = this.GetWatchOnlyWallet();
+
+            foreach (var address in watchOnlyWallet.WatchedAddresses.Values)
+            {
+                foreach (var transaction in address.Transactions.Values)
+                {
+                    lookup.Add(transaction.Id, transaction);
+                }
+            }
+
+            foreach (var transaction in watchOnlyWallet.WatchedTransactions.Values)
+            {
+                // It is conceivable that a transaction could be both watched
+                // in isolation and watched as a transaction under one or
+                // more watched addresses.
+                if (!lookup.TryAdd(transaction.Id, transaction))
+                {
+                    // Check to see if there is better information in
+                    // the watched transaction than the watched address.
+                    // If there is, use the watched transaction instead.
+
+                    var existingTx = lookup[transaction.Id];
+
+                    if ((existingTx.MerkleProof == null && transaction.MerkleProof != null) ||
+                        (existingTx.BlockHash == null && transaction.BlockHash != null))
+                    {
+                        lookup.Remove(transaction.Id);
```

Comment:
you don't need to remove it from the lookup.
If you update the `existingTx` it will be updated in the list automatically.
I suggest adding some unit tests to this method.